File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/presenters/login/PersonalSignUpPresenter.java
package app.editors.manager.mvp.presenters.login;

import java.util.Locale;

import app.editors.manager.R;
import app.editors.manager.app.Api;
import app.editors.manager.app.App;
import app.editors.manager.managers.exceptions.UrlSyntaxMistake;
import app.editors.manager.mvp.models.request.RequestRegister;
import app.editors.manager.mvp.models.response.ResponseRegisterPersonalPortal;
import app.editors.manager.mvp.views.login.PersonalRegisterView;
import moxy.InjectViewState;
import retrofit2.Response;

@InjectViewState
public class PersonalSignUpPresenter extends BaseLoginPresenter<PersonalRegisterView, ResponseRegisterPersonalPortal> {

    public static final String TAG = PersonalSignUpPresenter.class.getSimpleName();
    private static final String EMAIL_CODE = "201";

    public PersonalSignUpPresenter() {
        App.getApp().getAppComponent().inject(this);
    }

    public void registerPortal(final String email) {
        try {
            initRetrofitPref(Api.PERSONAL_HOST);
            final RequestRegister requestRegister = new RequestRegister();
            requestRegister.setEmail(email);
            requestRegister.setLanguage(Locale.getDefault().getLanguage());
            mRequestCall = mRetrofitTool.getApiWithPreferences().registerPersonalPortal(requestRegister);
            mRequestCall.enqueue(new BaseCallback() {

                @Override
                public void onSuccessResponse(Response<ResponseRegisterPersonalPortal> response) {
                    final String message = response.body().getResponse();
                    if (message != null && !message.isEmpty() && !response.body().getStatus().equals(EMAIL_CODE)) {
                        getViewState().onError(mContext.getString(R.string.errors_email_already_registered));
                        return;
                    } else if (message != null && !message.isEmpty()) {
                        getViewState().onError(message);
                        return;
                    }

                    getViewState().onRegisterPortal();
                }
            });
        } catch (UrlSyntaxMistake e) {
            // No need handle
        }
    }

}

File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/presenters/login/PersonalSignUpPresenter.kt
package app.editors.manager.mvp.presenters.login

import app.documents.core.model.cloud.CloudPortal
import app.documents.core.network.common.NetworkResult
import app.documents.core.network.common.contracts.ApiContract
import app.editors.manager.R
import app.editors.manager.app.App
import app.editors.manager.mvp.views.login.PersonalRegisterView
import kotlinx.coroutines.launch
import lib.toolkit.base.BuildConfig
import lib.toolkit.base.managers.utils.StringUtils
import moxy.InjectViewState
import moxy.presenterScope
import java.util.Locale

@InjectViewState
class PersonalSignUpPresenter : BaseLoginPresenter<PersonalRegisterView>() {

    companion object {
        val TAG: String = PersonalSignUpPresenter::class.java.simpleName
        private const val EMAIL_CODE = "201"

        const val TAG_INFO = "/#info"
    }

    init {
        App.getApp().appComponent.inject(this)
    }

    fun checkMail(email: String) {
        var isInfo = false
        var mail = email

        if (email.contains(TAG_INFO)) {
            isInfo = true
            mail = email.replace(TAG_INFO, "")
        }
        if (StringUtils.isEmailValid(mail.trim())) {
            registerPortal(mail, isInfo = isInfo)
        } else {
            viewState.onMessage(R.string.errors_email_syntax_error)
        }
    }

    private fun registerPortal(email: String?, isInfo: Boolean = false) {
        if (email == null || email.isEmpty() && !StringUtils.isEmailValid(email)) {
            viewState.onError(context.getString(R.string.errors_email_syntax_error))
            return
        }

        val url = if (isInfo) "${BuildConfig.SUBDOMAIN}.${BuildConfig.DEFAULT_INFO_HOST}" else ApiContract.PERSONAL_HOST
        App.getApp().refreshLoginComponent(CloudPortal(url = url))
        signInJob = presenterScope.launch {
            loginRepository.registerPersonal(email, Locale.getDefault().language)
                .collect { result ->
                    when (result) {
                        is NetworkResult.Success -> viewState.onRegisterPortal()
                        is NetworkResult.Error -> fetchError(result.exception)
                        is NetworkResult.Loading -> Unit
                    }
                }
        }
    }

}